src/parquet/arrow/arrow-reader-writer-test.cc

```diff
@@ -1104,61 +1100,314 @@ class TestNestedSchemaRead : public ::testing::Test {
     writer_->Close();
   }
 
-  void CreateNestedParquet() {
+  void MakeValues(int num_rows) {
+    std::shared_ptr<Array> arr;
+    ASSERT_OK(NullableArray<::arrow::Int32Type>(num_rows, 0, kDefaultSeed, &arr));
+    values_array_ = std::dynamic_pointer_cast<::arrow::Int32Array>(arr);
+  }
+
+  void WriteColumnData(size_t num_rows, int16_t* def_levels,
+      int16_t* rep_levels, int32_t* values) {
+    auto typed_writer = static_cast<TypedColumnWriter<Int32Type>*>(
+      row_group_writer_->NextColumn());
+    typed_writer->WriteBatch(num_rows, def_levels, rep_levels, values);
+  }
+
+  void ValidateArray(std::shared_ptr<Array> array, size_t expected_nulls) {
+    ASSERT_EQ(array->length(), values_array_->length());
+    ASSERT_EQ(array->null_count(), expected_nulls);
+    // Also independently count the nulls
+    auto local_null_count = 0;
+    for (int i = 0; i < array->length(); i++) {
+      if (array->IsNull(i)) {
+        local_null_count++;
+      }
+    }
+    ASSERT_EQ(local_null_count, expected_nulls);
+  }
+
+  void ValidateColumnArray(std::shared_ptr<::arrow::Int32Array> array,
+      size_t expected_nulls) {
+    ValidateArray(array, expected_nulls);
+
+    int j = 0;
+    for (int i = 0; i < values_array_->length(); i++) {
+      if (array->IsNull(i)) {
+        continue;
+      }
+      ASSERT_EQ(array->Value(i), values_array_->Value(j));
+      j++;
+    }
+  }
+
+  // A parquet with a simple nested schema
+  void CreateSimpleNestedParquet(Repetition::type struct_repetition) {
     std::vector<NodePtr> parquet_fields;
-    std::shared_ptr<Array> values;
+    // TODO(itaiin): We are using parquet low-level file api to create the nested parquet
+    // this needs to change when a nested writes are implemented
 
     // create the schema:
-    // required group group1 {
+    // <struct_repetition> group group1 {
     //   required int32 leaf1;
-    //   required int32 leaf2;
+    //   optional int32 leaf2;
     // }
     // required int32 leaf3;
 
-    parquet_fields.push_back(GroupNode::Make("group1", Repetition::REQUIRED,
+    parquet_fields.push_back(GroupNode::Make("group1", struct_repetition,
         {PrimitiveNode::Make("leaf1", Repetition::REQUIRED, ParquetType::INT32),
-            PrimitiveNode::Make("leaf2", Repetition::REQUIRED, ParquetType::INT32)}));
+         PrimitiveNode::Make("leaf2", Repetition::OPTIONAL, ParquetType::INT32)}));
     parquet_fields.push_back(
         PrimitiveNode::Make("leaf3", Repetition::REQUIRED, ParquetType::INT32));
 
-    const int num_columns = 3;
     auto schema_node = GroupNode::Make("schema", Repetition::REQUIRED, parquet_fields);
 
-    InitNewParquetFile(std::static_pointer_cast<GroupNode>(schema_node), 0);
+    // Create definition levels for the different columns that contain interleaved
+    // nulls and values at all nesting levels
```

Comment:
Using JSON and fuzzing sounds great -  would make testing less of a pain and would definitely find way more bugs (e.g. the one I found yesterday..). 
I do believe that working on writing should happen right after solving full nested data reads.